File: src/structure/package/packageToFileName.ts
import path from "node:path";
import {Configuration} from "../../configuration/configuration.js";
import {moduleNameToPackage} from "../module/moduleNameToPackage.js";
import {removePrefix} from "../removePrefix.js";

export function dirNameToPackage(dirName: string) {
    return dirName
        .split(path.posix.sep)
        .filter(it => it !== "")
}

export function packageToFileName(
    packageChunks: string[],
    fileName: string,
) {
    return path.posix.join(
        ...packageChunks,
        fileName,
    )
}

export function packageToOutputFileName(
    packageChunks: string[],
    fileName: string,
    configuration: Configuration,
) {
    const {libraryName, libraryNameOutputPrefix} = configuration

    const result = packageToFileName(packageChunks, fileName)

    if (libraryNameOutputPrefix) {
        return result
    }

    const basePackage = path.posix.join(...moduleNameToPackage(libraryName)) + path.posix.sep

    return removePrefix(result, [basePackage])
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/package/packageToFileName.kt
package io.github.sgrishchenko.karakum.structure.`package`

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.structure.module.moduleNameToPackage
import io.github.sgrishchenko.karakum.structure.removePrefix
import js.array.ReadonlyArray
import node.path.path

fun dirNameToPackage(dirName: String): ReadonlyArray<String> {
    return dirName
        .split(path.posix.sep)
        .filter { it.isNotEmpty() }
        .toTypedArray()
}

fun packageToFileName(
    packageChunks: ReadonlyArray<String>,
    fileName: String,
): String {
    return path.posix.join(
        *packageChunks,
        fileName,
    )
}

fun packageToOutputFileName(
    packageChunks: ReadonlyArray<String>,
    fileName: String,
    configuration: Configuration,
): String {
    val libraryName = configuration.libraryName
    val libraryNameOutputPrefix = configuration.libraryNameOutputPrefix

    val result = packageToFileName(packageChunks, fileName)

    if (libraryNameOutputPrefix) {
        return result
    }

    val basePackage = path.posix.join(paths = moduleNameToPackage(libraryName)) + path.posix.sep

    return removePrefix(result, arrayOf(basePackage))
}
